### ISSUE TITLE
ConvexHull generation from certain input points causing infinite loop

### ISSUE THREAD

**Issue by Glidias** [Mugen87/yuka#11]
This is a showstopper for me. I'm not sure where the loop goes endless from the YUKA.ConvexHull codebase yet. Example point input below for reference.

It'd be good to have a test unit/case for this.

To trigger endless loop, use the code below:.
```
var jsonArr = [{"x":-55.53499860610813,"y":73.11029657847513,"z":-63.59099818190188},{"x":-55.7849986005202,"y":73.11029657847513,"z":-69.34100186807655},{"x":-58.53499853905291,"y":73.11029657847513,"z":-68.84100187925242},{"x":-58.03499855022878,"y":73.11029657847513,"z":-63.59099818190188},{"x":-55.53499860610813,"y":73.01029657847513,"z":-63.59099818190188},{"x":-55.7849986005202,"y":73.01029657847513,"z":-69.34100186807655},{"x":-58.53499853905291,"y":73.01029657847513,"z":-68.84100187925242},{"x":-58.03499855022878,"y":73.01029657847513,"z":-63.59099818190188},{"x":-9.78499962870032,"y":74.11029655612339,"z":-64.84100196865938},{"x":-10.034999623112384,"y":74.11029655612339,"z":-64.09100198542319},{"x":-9.034999645464126,"y":73.92279656031434,"z":-63.59099818190188},{"x":-9.78499962870032,"y":74.0102965561234,"z":-64.84100196865938},{"x":-10.034999623112384,"y":74.0102965561234,"z":-64.09100198542319},{"x":-9.034999645464126,"y":73.82279656031434,"z":-63.59099818190188}];
new YUKA.ConvexHull().fromPoints(jsonArr.map((p)=>{return new YUKA.Vector3(p.x, p.y, p.z)}));
  ```

Screenshot of the above 14 input points below (shown as small green cubes) floating above the navmesh regions highlighted in red. Looks "innocent" enough. So this seems like a serious bug.

![image](https://user-images.githubusercontent.com/190195/68199237-9b0fff80-fff8-11e9-822c-0761871e6829.png)

expected hull output:

![image](https://user-images.githubusercontent.com/190195/68205159-89345980-0004-11ea-9c4c-3778013d0b77.png)

**Comment by Mugen87** [comment]
I can reproduce. It looks like this happens in `ConvexHull._mergeFaces()`. Investigating...

**Comment by Mugen87** [comment]
I've provided a fix in a separate branch: a2cbee5b3653bb61d55a497ef339eec56546a351

Can you also test this build in your project? It was quite complicated to understand what's going wrong and I'm still considering a different type of solution. However, the commit should be okay as a first fix.

**Comment by Glidias** [comment]
Appears to work fine. 2 cases out of several where `_mergePossible(` filters out. I'm not sure how that case could happen though visually, since if that were true wouldnt one of the polygons be no longer be convex  to end up having multiple edges into the same polygon? THe funny thing is if you run the same set of faces (non-merged) through NavMesh, there is no infinite loop. The loop happens when it appears that after some merging, there is no more loop back to starting polygon.edge through the linked list thus the `polygon.convex()` check method hangs.

On a sidenote, merging faces could be an optional thing, right? I wonder is skipping the need to merge faces and testing the intersections and other methods with non-merged faces is viable (i think it is?), and whether, on the whole, skipping the merging faces is better for performance as well. If it's not going to be frequent multiple tests with the same convex hull, but only a one-off case, i wonder if just sticking to non-merged faces would be more performant overall even as the sample size is marginally larger.

**Comment by Mugen87** [comment]
> On a sidenote, merging faces could be an optional thing, right?

Yes, but I do think that a convex hull with less faces is in general preferable. In our previous scenarios, we usually generate the convex hull once and then use it for intersection tests. The code is actually not optimized if you generate hulls over and over again.

The merge code in `ConvexHull` and `NavMesh` is more or less equal and will fail if it encounters the above test case. Unfortunately, the merge can produce a polygon which edges are not correctly linked. So it's not possible to use `edge.next` to move along the contour. But only if both original convex polygons are already connected. I'm still trying to understand what that means and if it is possibly a degenerated case produced elsewhere in the code.

**Comment by Mugen87** [comment]
bbf86d08fdf357b18a915c8a3c06403f95e1da85 enhanced the unit tests and also added `ConvexHull.mergeFaces` (default `true`). You can define whether merging is desired or not.

@Glidias Is this helpful to you?

**Comment by Mugen87** [comment]
Okay. In any event, having a flag that controls the merge might be useful in the future (if only for debugging^^).

**Comment by Glidias** [comment]
Regarding your mergeFaces check, woudln't it skip the remaining lines in `_mergeFaces()` which are important?

```
// recompute centroid of faces

		for ( let i = 0, l = faces.length; i < l; i ++ ) {

			faces[ i ].computeCentroid();

		}

		// compute centroid of convex hull and the final edge and vertex list

		this.computeCentroid();
		this.computeUniqueEdges();
		this.computeUniqueVertices();
```

**Comment by Mugen87** [comment]
> Regarding your mergeFaces check, woudln't it skip the remaining lines in _mergeFaces() which are important?

Good point. Refactored the code a little bit so these lines are not skipped anymore.

> If you already know beforehand that a set of polygons, forms a convex hull, is it possible to form a ConvexHull from existing faces skipping the fromPoints generation part?

What is the use case of doing this? Would the method look like so?

```js
fromFaces( faces ) {

    this.faces = faces;

    this.computeCentroid();
    this.computeUniqueEdges();
    this.computeUniqueVertices();

    return this;

}
```

**Comment by Mugen87** [comment]
The original issue should be resolved via 7fd0d290279a8e99f21e805aa90cfd2fd49e36d6.
